src/type_object.rs

```diff
@@ -132,78 +136,65 @@ where
     T: PyTypeInfo,
 {
     fn type_object(py: Python) -> &PyType {
-        unsafe { py.from_borrowed_ptr(<Self as PyTypeInfo>::type_object() as *const _ as _) }
+        unsafe { py.from_borrowed_ptr(Self::type_object_raw(py) as *const _ as _) }
     }
 }
 
-/// Lazy type object for Exceptions
-#[doc(hidden)]
-pub struct LazyHeapType {
-    value: UnsafeCell<Option<NonNull<ffi::PyTypeObject>>>,
-    initialized: AtomicBool,
-}
-
-impl LazyHeapType {
-    pub const fn new() -> Self {
-        LazyHeapType {
-            value: UnsafeCell::new(None),
-            initialized: AtomicBool::new(false),
-        }
-    }
-
-    pub fn get_or_init<F>(&self, constructor: F) -> NonNull<ffi::PyTypeObject>
-    where
-        F: Fn(Python) -> NonNull<ffi::PyTypeObject>,
-    {
-        if !self
-            .initialized
-            .compare_and_swap(false, true, Ordering::Acquire)
-        {
-            // We have to get the GIL before setting the value to the global!!!
-            let gil = Python::acquire_gil();
-            unsafe {
-                *self.value.get() = Some(constructor(gil.python()));
-            }
-        }
-        unsafe { (*self.value.get()).unwrap() }
-    }
-}
-
-// This is necessary for making static `LazyHeapType`s
-//
-// Type objects are shared between threads by the Python interpreter anyway, so it is no worse
-// to allow sharing on the Rust side too.
-unsafe impl Sync for LazyHeapType {}
-
 /// Lazy type object for PyClass
 #[doc(hidden)]
 pub struct LazyStaticType {
-    value: UnsafeCell<ffi::PyTypeObject>,
-    initialized: AtomicBool,
+    // Boxed because Python expects the type object to have a stable address.
+    value: OnceCell<Box<ffi::PyTypeObject>>,
```

Comment:
Why can't we use `ffi::PyTypeObject` here?
I think `&value` is a consistent address since `LazyStaticType` is only used as a static variable.

Reply:
`OnceCell::get_or_init` moves the result. I'm not sure that it's safe to do this after calling `PyType_Ready`, which is why I boxed it...

Reply:
> I'm not sure that it's safe to do this after calling PyType_Ready, which is why I boxed it...

Yeah, it's not safe. But then why we need to use `OnceCell` here? Just for simplification?

Reply:
I wanted to use `OnceCell` because then all our internal statics (datetime API, imported type objects, created type objects) are initialised the same way, using `OnceCell`. Changing to use `OnceCell` also helped us discover the unsafety in previous implementation's interaction with class attributes, so I think using `OnceCell` here is proving a good idea.

Reply:
Hmm I still don't like this redundant `Box, though I'm not strongly against it.

Reply:
How about I change it to `Pin<Box<ffi::PyTypeObject>>`? That way it clearly shows that it should have a fixed address.

Reply:
My concern is about the allocation cost.
`ffi::PyTypeObject` is (roughly) 320bytes.

Reply:
`Pin` looks nice, though.

Reply:
`Pin<P>` is not really about having a stable address, it’s about not being able to move out of the place pointed by `P`. `Box` already guarantees to have a stable address, I think `Pin` only adds confusion.

Reply:
Is thta not kinda the idea though - once `PyType_Ready` is called, the type object should not be moved in memory? (Hence why I thought `Pin` is applicable here.)

Reply:
You can still move in or out of the logical `ffi::PyTypeObject` without problem: it’s `Copy` and `Unpin` anyway (it basically amounts to overwriting all the fields with a `memcpy`).

It’s the place in memory where the `ffi::PyTypeObject` is allocated that should not move.

Reply:
That's a fair point. Ok I'm convinced adding `Pin` doesn't help here.

@kngwyu I don't think it's possible to remove the `Box` without adding some `unsafe` inplace initialization trick to `GILOnceCell`- and I'm not sure even then I can make it sound.

With regards to the allocation concern, I think this Python code will cause the same allocation cost anyway:

```
class Foo:
    pass
```

Reply:
> I'm not sure even then I can make it sound.

What we need is just having `GILOnceCell<()>` and `UnsafeCell<ffi::PyTypeObject>`, and then initializing the type object in `get_or_init`.

> With regards to the allocation concern, I think this Python code will cause the same allocation cost anyway:

Yeah, I think I can be *too nervous* about allocation. This is why I said ' I'm not strongly against it'.

Reply:
> What we need is just having `GILOnceCell<()>` and `UnsafeCell<ffi::PyTypeObject>`, and then initializing the type object in `get_or_init`.

Hmm I guess that can work. I think we would need to be careful to ensure thread safety and no deadlocks.

As it's an implementation detail, maybe we should focus on this in a new PR (perhaps the same one that also solves recursive class attributes problem)?

Reply:
@kngwyu’s idea is doable once we’re sure we don’t release the GIL during the initialization of the type object. So I agree with @davidhewitt that is linked to the class attributes problem (the only part that calls into user code).

Reply:
@davidhewitt @scalexm 
Thank you for your comments.

> As it's an implementation detail, maybe we should focus on this in a new PR (perhaps the same one that also solves recursive class attributes problem)?

Agreed :+1: